LD scores method

liameabbott opened the pull request:
Bundling up the LD score calculation work @jbloom22 and @ktashman did during the Hail hackathon into a Hail method. Not sure if this should be a part of core Hail methods, or sequestered as a user-contributed method, of which there will hopefully be more in 0.2. I've currently taken the latter route, but we can discuss at office hours tomorrow @tpoterba.

jbloom22 commented:
Cool! As we just discussed, there's a complication here for dealing with multiple contigs, and a question of enforcing ordering, that should be resolved by a PR with a locus_windows function that I hope to get in this week, which can then be incorporated here.

liameabbott commented:
I think it should deal with multiple contigs appropriately now. Still relies on consistent (ascending by position within contig) ordering between collected variants, MatrixTable rows, and BlockMatrix rows. Will wait for your locus_windows function to help clean this up.

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+@typecheck(entry_expr=expr_numeric,
+           annotation_exprs=oneof(expr_numeric, sequenceof(expr_numeric)),
+           position_expr=expr_numeric,
+           window_size=oneof(int, float))
```
suggestion: we have `numeric` as a typechecker for this

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+    Warning
+    -------
+    :func:`.ld_score` will fail if ``entry_expr`` results in any missing values. The special float value ``nan`` is not considered a missing value.
+
```
unnecessary extra space

liameabbott replied:
Is this referencing the double newline after the warning section? Without the extra newline, the "Further Reading" section is lumped into the warning section.

tpoterba replied:
Ah. The best thing to do then is probably tab over the line after the warning. Numpy docstrings have some weird rules.

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+
+    For more in-depth discussion of LD scores, see:
+
+    - `LD Score regression distinguishes confounding from polygenicity in genome-wide association studies (Bulik-Sullivan et al, 2015) <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4495769/>`__
```
in our main codebase, we try to keep lines a bit shorter. I wrote a tool to automate this by block:

https://hail.is/tools/formatter.html

liameabbott replied:
I'm not sure how to split back-ticked phrases across multiple lines.

tpoterba replied:
I think it'll work as expected, the link will stay. We can check by looking at the docs when they build

liameabbott replied:
I'm getting build errors trying to split the lines. Happy to change this later if we figure out how.

tpoterba replied:
ok, dont worry about it

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+    contig = '0'
+    
+    for i, (c, p) in enumerate(positions):
+
```
more stylistic to delete these newlines

tpoterba commented on python/hail/tests/test_contrib.py:
```diff
@@ -0,0 +1,54 @@
+        ht = ht.annotate(locus=hl.locus(ht.CHR, ht.BP))
+        ht = ht.key_by('locus')
+
+        mt = hl.import_plink(bed=resource('ldsc.bed'), bim=resource('ldsc.bim'), fam=resource('ldsc.fam'))
```
this comment I do feel strongly about.

liameabbott replied:
Nice, was wondering if something like that existed.

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+
+    Returns
+    -------
+    :class:`.HailTable` 
```
s/HailTable/Table

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+
+    Parameters
+    ----------
+    entry_expr  : :class:`.NumericExpression`
```
unnecessary double space in a few params here

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+           window_size=oneof(int, float))
+def ld_score(entry_expr, annotation_exprs, position_expr, window_size) -> Table:
+    """
+    Calculate LD scores.
```
convention is to put the one-liner on the same line as the triple quotes

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+    >>> ht = ht.key_by('locus')
+    
+    >>> # Annotate MatrixTable with external annotations 
+    >>> mt = mt.annotate_rows(univariate_annotation=hl.literal(1),   # univariate LD scores
```
hl.literal(1) can just be `1`

tpoterba commented on python/hail/contrib/ldsc.py:
```diff
@@ -0,0 +1,156 @@
+
```
I'm a bit worried about the name clash from calling this `ldsc`. maybe `ldscore` to make it totally clear?

liameabbott commented:
I rebased and addressed most of your comments. If this looks good enough to you Tim, maybe we can merge it in to experimental for now, and when @jbloom22's locus_window function is in I can work on incorporating that into ld_score before promoting from experimental to Hail methods.

tpoterba commented on python/hail/tests/test_contrib.py:
```diff
@@ -0,0 +1,54 @@
+import unittest
```
delete this file

tpoterba commented on python/hail/__init__.py:
```diff
@@ -27,6 +27,7 @@
 from .expr import *
 from .genetics import *
 from .methods import *
+from .experimental import *
```
delete

liameabbott replied:
I would like to be able to call `hl.ld_score()`, not `hl.experimental.ld_score()`.

tpoterba replied:
I think experimental stuff needs to be explicitly such. @cseed thoughts?

liameabbott replied:
I'll just change it for now, but if there are going to be methods that sit in experimental long-term rather than a short stay en route to becoming a core method, I think users would prefer the former for regular use.
